GjallarhornTest2/Program.fs

```diff
@@ -38,59 +37,91 @@ module Program =
            "Pollux"
            "Avior"
         ]
+    
+    type ItemMessage = UpdateItem
+
+    [<CLIMutable>]
+    type Item =
+        {
+            Name : string
+            Top : float
+            Left : float
+        }
+
+    let createItem name = { Name = name; Top = 200.; Left = 200. }
+
+    [<CLIMutable>]
+    type ItemViewModel =
+        {
+            XName : string
+            Top : float
+            Left : float
+            Self : ISignal<Item>
+        }
 
-    type Msg =
+    let designSignal = createItem "" |> Signal.constant
+    let itemd = { XName = ""; Top = 0.; Left = 0.; Self = designSignal }
+
+    let itemComponent =
+        let update msg itemModel =
+            match msg with
+            | UpdateItem -> itemModel
+
+        Component.create<Item, unit, _>
+            [
+                <@ itemd.XName @> |> Bind.oneWay (fun item -> item.Name)
+                <@ itemd.Top @> |> Bind.oneWay (fun item -> item.Top)
+                <@ itemd.Left @> |> Bind.oneWay (fun item -> item.Left)
+                <@ itemd.Self @> |> Bind.self
+            ]
+        |> Component.toSelfUpdating update
+
+    type ModelMessage =
         | Add of string
 
-    type Thing =
+    type Model =
         {
             Source : string list
-            Target : string list
+            Target : Item list
         }
     with
         static member Default = { Source = names; Target = [] }
-    
-    let update msg (thing : Thing) =
-        match msg with
-        | Add item ->
-            if thing.Target |> List.contains item
-            then thing
-            else { thing with Target = (item :: thing.Target) |> List.sort }
-
-    
-    type ItemViewModel =
-        {
-            Name : string
-            Self : ISignal<string>
-        }
 
+    let updateModel msg (model : Model) =
+        match msg with
+        | Add name ->
+            if model.Target |> List.exists (fun item -> item.Name = name)
+            then model
+            else
+                { model with
+                    Target = (createItem name :: model.Target) |> List.sort }
 
     [<CLIMutable>]
     type ViewModel =
         {
-            Source : string list
+            XSource : string list
             Add : VmCmd<string>
-            Target : string list
+            XTarget : ItemViewModel list
```

Comment:
I intentionally changed the name of the properties in the ViewModel to be different from the model, so I could be sure the XAML binds to the correct thing, but while the bindings work, the `ItemsSource` in the running application is a list of `Item`, and I don't understand how that happens.

Reply:
That happens because line 115 and 116 map "m.Target" (which is `Item list` - line 115) to be bound under the name "XTarget" (line 116)

Reply:
That happens because line 115 and 116 map "m.Target" (which is `Item list` - line 115) to be bound under the name "XTarget" (line 116).

See https://github.com/TeaDrivenDev/GjallarhornTest2/pull/1/files#diff-9e5440c33ff507e2cab450df425fcf2cR115

Reply:
But why does it bind to the `Model` type at all? From all I seemed to understand, `ViewModel` is what it's supposed to bind to. I apparently still miss something very fundamental to the whole concept.

Reply:
The "ViewModel" is not bound... It's a completely optional thing used only for avoiding magic strings and helping xaml hints (ie: design time data). It binds via the map call I linked

Reply:
Okay, that is a really crucial piece of the puzzle that had eluded me so far. ;-)

Reply:
Think of it purely as design data.. the schema of what you'll bind. Setting as design data fixes mutable issues, let's you use quotations instead of strings, etc. At runtime, it's unused

Reply:
Got that, and things make a lot more sense now. Now, where do I plug that component for `Item`? That seems to belong somewhere "inside" the mapping code for `Model`.

Reply:
That's where collectionbinding comes in. You give it a mapping to a `'a list` (or any seq) and a component to use per item. It automatically wires up everything. Think of it like having a list of DTO and a Func<DtoVM> in c# terms

Reply:
I see, but how? I tried replacing the `Bind.Explicit.oneWay` call at https://github.com/TeaDrivenDev/GjallarhornTest2/blob/0ff3f7ee13751a24dd321800a68a7609cfd68cad/GjallarhornTest2/Program.fs#L119 with `Bind.collection`, but that takes a `Dispatch<'Nav>`, and I'm at a loss as to what that means, because there is just nothing in there that I'd consider "navigation".

Reply:
The `_nav` param to that function... just pass it through...

Reply:
FYI - `_nav` there is a function that allows you to push out a "navigation message" - in your case, you aren't using the nav system (single view from window is like a SPA ), but it's how you push out requests to navigate. The collection binfijng needs it so it can pass through to the individual item components, I'm in case they wanted to request a change in navigation (like open a dialog to edit something, create a report, etc)

Reply:
I suppose that's supposed to go in the list at https://github.com/TeaDrivenDev/GjallarhornTest2/blob/0ff3f7ee13751a24dd321800a68a7609cfd68cad/GjallarhornTest2/Program.fs#L123 then? The problem there is that it's an `IObservable<'Msg> option` for some reason, while in that list I'd need just an `IObservable<'Msg>`, so there seems to be something still missing there.

Reply:
What are you trying to have happen? I can play with it. (It's all running now - so just wondering what you want to happen to the items...)

Reply:
I want the Items to update their Top and Left coordinates in the model when they're dragged around.

Reply:
Okay - see https://github.com/TeaDrivenDev/GjallarhornTest2/pull/2